src/librustc_mir/transform/add_yk_swt_calls.rs

```diff
@@ -131,51 +131,54 @@ impl MirPass for AddYkSWTCalls {
     }
 }
 
-/// Given a `MirSource`, decides if we should annotate the correpsonding MIR.
-fn should_annotate(tcx: TyCtxt<'a, 'tcx, 'tcx>, src: MirSource) -> bool {
-    // Never annotate any MIR-like thing marked `#[no_trace]` or `#[naked]`. The trace record and
-    // wrapper are also marked `#[no_trace]` to prevent infinite recursion.
+/// Given a `MirSource`, decides if it is possible for us to trace (and thus whether we should
+/// transform) the MIR. Returns `true` if we cannot trace, otherwise `false`.
+fn is_untraceable(tcx: TyCtxt<'a, 'tcx, 'tcx>, src: MirSource) -> bool {
+    // Never annotate anything annotated with the `#[no_trace]` attribute. This is used on tests
+    // where our pass would interfere and on the trace recorder to prevent infinite
+    // recursion.
+    //
+    // "naked functions" can't be traced because their implementations manually implement
+    // binary-level function epilogues and prologues, often using in-line assembler. We can't
+    // automatically insert our calls into such code without breaking stuff.
     for attr in tcx.get_attrs(src.def_id).iter() {
         if attr.check_name("no_trace") {
-            return false;
+            return true;
         }
         if attr.check_name("naked") {
-            return false;
+            return true;
        }
     }
 
-    // If there is a crate level `#![no_trace]` attribute, honour that.
+    // Similar to `#[no_trace]`, don't transform anything inside a crate marked `#![no_trace]`.
     for attr in tcx.hir.krate_attrs() {
         if attr.check_name("no_trace") {
-            return false;
+            return true;
         }
     }
 
-    // We can't call the software tracing function if there is no libcore.
+    // We can't call the software tracing function if there is no libcore because that's where the
+    // entry point to the trace recorder function lives.
     if attr::contains_name(tcx.hir.krate_attrs(), "no_core") {
-        return false;
+        return true;
     }
 
     // The libcompiler_builtins crate is special and we can't annotate it.
```

Comment:
To be honest, the compiler_builtins crate is a bit of a mystery to me. All I know is that strange things happen if you do transform it.

Ah: s/annotate/transform/ too.

If you like, we can remove this check, run bors, and observe the error message for clues as to what to put in this comment? I don't recall exactly what happens off the top of my head.

Reply:
It might be a good idea. If we've already forgotten, someone in the future's going to worry about it too. Since this is such a critical part of the whole infrastructure, I suspect our future selves will be grateful for such hints :)

Reply:
Okie dokie! Here goes then.